source/common/router/header_parser.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <list>
+#include <string>
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/http/header_map.h"
+
+#include "common/protobuf/protobuf.h"
+#include "common/router/header_formatter.h"
+
+#include "api/base.pb.h"
+
+namespace Envoy {
+namespace Router {
+
+class RequestHeaderParser;
+typedef std::unique_ptr<RequestHeaderParser> RequestHeaderParserPtr;
+
+/**
+ * This class provides request-time generation of upstream request headers. Header configurations
+ * are pre-parsed to select between constant values and values based on the evaluation of
+ * AccessLog::RequestInfo fields.
+ */
+class RequestHeaderParser {
+public:
+  virtual ~RequestHeaderParser() {}
+
+  static RequestHeaderParserPtr
+  parse(const Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>& headers);
+
+  void evaluateRequestHeaders(Http::HeaderMap& headers,
+                              const AccessLog::RequestInfo& request_info) const;
+
+private:
+  std::list<std::pair<Http::LowerCaseString, HeaderFormatterPtr>> header_formatters_;
```

Comment:
I vector is faster to iterate over than a list; I think it is a better choice because nothing is ever added/removed after this is initialized, right?

Reply:
These only moved, but I'll switch them to vector.

Reply:
(This will now come in another PR.)